src/components/ControlledForm/index.tsx

```diff
@@ -0,0 +1,85 @@
+import React, { useState } from 'react';
+import { ControlledInputProps } from '../Input';
+
+export interface Field {
+  name: string;
+  value?: any;
+  isValid?: boolean;
+}
+
+type Props = {
+  initialValues?: object;
+  onSubmit: (values) => void;
+  render: (formState: {
+    isValid: boolean;
+    state: { [name: string]: Field };
+    getInputControls: <T = string>(name: string) => ControlledInputProps<T>;
+  }) => React.ReactNode;
+};
+
+const ControlledForm = (props: Props) => {
+  const [fields, setFields] = useState<{ [key: string]: Field }>({});
+
+  const isValid = !Object.values(fields).filter(
+    field => field.isValid !== undefined && !field.isValid
+  ).length;
+
+  const onSubmitHandler = formEvent => {
+    formEvent.preventDefault();
+
+    const values = Object.keys(fields).reduce((previous, current) => {
+      previous[current] = fields[current].value;
+      return previous;
+    }, {});
+
+    if (isValid) {
+      props.onSubmit(values);
+    }
+  };
+
+  const onInputValueChange = <T,>(field: string, value?: T) => {
+    setFields(f => (f[field] ? { ...f, [field]: { ...f[field], value } } : f));
+  };
+
+  const onValidationChange = (field: string, isValid: boolean) => {
+    setFields(f => (f[field] ? { ...f, [field]: { ...f[field], isValid } } : f));
+  };
+
+  const getFieldValue = <T,>(field: string): T => {
+    return fields[field]?.value;
```

Comment:
if there is no value for a given field, this function will return undefined, and not `T`. I recommend adding undefined as a possible return type